File: 0743-network-delay-time/0743-network-delay-time.cpp
const int INF = numeric_limits<int>::max();
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<pair<int, int>>> graph(n + 1);
        vector<int> dist(n + 1, INF);
        dist[k] = 0;

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.emplace(0, k);

        // build graph
        for (const auto &time : times) {
            graph[time[0]].emplace_back(time[1], time[2]);
        }

        // dijkstra's algo
        while (!pq.empty()) {
            const auto [dis, node] = pq.top();
            pq.pop();

            if (dis > dist[node]) {
                continue;
            }

            for (const auto [neighbor, weight] : graph[node]) {
                if (dist[node] + weight < dist[neighbor]) {
                    dist[neighbor] = dist[node] + weight;
                    pq.emplace(dist[neighbor], neighbor);
                }
            }
        }

        return any_of(dist.begin() + 1, dist.end(), [](const int x) {
            return x == INF;
        }) ? -1 : *max_element(dist.begin() + 1, dist.end());
    }
};

File: 0743-network-delay-time/0743-network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # directed graph with edge value node 1 to n
        # k start node
        # return the max shotest path from k to all other node if all node is visgtisted
        # otherwise return -1
        # ti is non negative
        # dijkstra to find single source shotest path
        # time complexity o((V + E)*logv) = o((n + len(times))*logn)
        graph = [[] for _ in range(n + 1)]
        for u, v, t in times:
            graph[u].append((v, t))

        total_time = [float("inf") for _ in range(n + 1)]
        total_time[k] = 0
        min_delay_time = 0
        pq = [(total_time[k], k)]

        while pq:
            time, cur = heapq.heappop(pq)
            if total_time[cur] < time:
                continue 
            
            min_delay_time = max(min_delay_time, time) 

            for nbr, nbr_time in graph[cur]:
                if time + nbr_time < total_time[nbr]:
                    total_time[nbr] = time + nbr_time
                    heapq.heappush(pq, (total_time[nbr], nbr))

        return min_delay_time if all(time != float("inf") for time in total_time[1:]) else -1




        